Episode multiple versions

SenorSmartyPants opened the pull request:
<!--
Ensure your title is short, descriptive, and in the imperative mood (Fix X, Change Y, instead of Fixed X, Changed Y).
For a good inspiration of what to write in commit messages and PRs please review https://chris.beams.io/posts/git-commit/ and our documentation.
-->

**Changes**
Enable multiple versions for episodes. Episode suffixes work or episode may be placed in it's own directory. Will not collapse episodes in series directory. Unit tests included.

```
/TV/Dexter/Dexter - S01E01/Dexter - S01E01 - One.mkv
/TV/Dexter/Dexter - S01E01/Dexter - S01E01 - Two.mkv
/TV/Dexter/Dexter - S01E02 - Another.mkv
/TV/Dexter/Dexter - S01E03 - Again.mkv
/TV/Dexter/Season 2/Dexter - S02E01/Dexter - S02E01 - First.mkv
/TV/Dexter/Season 2/Dexter - S02E01/Dexter - S02E01 - Second.mkv
/TV/Dexter/Season 2/Dexter - S02E02 - More.mkv
```

JF will show 3 episodes in season 1. 2 episodes in season 2.

SenorSmartyPants committed:
Non greedy regex, parse the first year, not the last

SenorSmartyPants commented:
One change that could affect current users. Movie versions that include a year as part of the version name need to be placed in brackets. Otherwise they appear as a possible episode and won't merge with other versions.

SenorSmartyPants commented:
This would address https://features.jellyfin.org/posts/251/multiple-versions-drop-down-on-tv-series

SenorSmartyPants commented:
Related discussion. https://github.com/jellyfin/jellyfin/discussions/7900

almightiest commented:
I like these being in a separate folder, so it may not conflict with: https://features.jellyfin.org/posts/198/handle-movies-with-multiple-split-files-as-one-item

But there may be opportunity to tackle both here eventually?  Maybe detect if it's "Part 1" "Part 2" (or some other naming convention for multi parts), and then be sure if you choose Part 1 to play, that Part 2 would auto play in sequential order?  Worth it to have any tests around this type of naming?

Some examples from my library (Which works in Plex/Kodi but not Jellyfin):

```
Armchair.Thriller/Armchair.Thriller.S02/s02e29.Part1-armchair.thriller.chelsea.murders.feature-cd1.avi
Armchair.Thriller/Armchair.Thriller.S02/s02e29.Part2-armchair.thriller.chelsea.murders.feature-cd2.avi

The.Hollow.Crown/The.Hollow.Crown.S01/the.hollow.crown.s01e01.part1.dvdrip.xvid.cd1.avi
The.Hollow.Crown/The.Hollow.Crown.S01/the.hollow.crown.s01e01.part2.dvdrip.xvid.cd2.avi
```

SenorSmartyPants commented:
Latest commit adds support for multiversion episodes in mixed folders (like season and series) so this file list is now supported. (Which Emby supports)

```
/TV/Dexter/Dexter - S01E01 - One.mkv
/TV/Dexter/Dexter - S01E01 - One plus.mkv
/TV/Dexter/Dexter - S01E02 - Two.mkv
/TV/Dexter/Dexter - S01E03 - Three.mkv
/TV/Dexter/Dexter - S02E01 - Ia.mkv
/TV/Dexter/Dexter - S02E01 - I.mkv
/TV/Dexter/Dexter - S02E02.mkv
```

SenorSmartyPants commented:
Force pushed to rebase to current master.

JaidenW commented:
Super interested in this pull.

Question, in it's current format, would this allow me to have multiple versions of an episode, even if they are stored in different directories? or would they all have to be combined within one folder? What i'm hoping to achieve is to store the majority of my content on one server, while dedicating another to strictly UHD/4k content.

Thanks!

SenorSmartyPants commented:
Unfortunately no. The episodes would have to be in the same folder to be merged.

JaidenW commented:
> Unfortunately no. The episodes would have to be in the same folder to be merged.

Alas, perhaps some day. Keep up the good work though!

SenorSmartyPants commented:
> 

Lots of code would have to be changed to get this to work. 
Check out mergerfs or another union file system. Maybe you can make it look like they are in the same directory to jellyfin.

ireun commented:
Also, this could probably be helpful in implementing this: https://features.jellyfin.org/posts/284/convert-option-similar-to-plexs-optimise

JeWe37 commented:
What is the state of this now(I have a structure like @JaidenW , but I can work around that with a simple overlayfs mount, I don't need it to be writable)? It seems there is a lot of debate about how to best do the parsing and this PR currently seems to have gone dead? Anything preventing me from just rebasing onto master and building from source with this PR applied?

SenorSmartyPants commented:
This PR is waiting on reviewers from the JF team. 

Nothing to stop you merging in your local copy.

JeWe37 commented:
So I've been trying this out and have not been able to get it to work no matter what I tried. I use Season folders and a structure like 
```
Name (2020)/Season 1/Name (year) - S01E01.mkv
Name (2020)/Season 1/Name (year) - S01E01 - [VERSION].mkv
Name (2020)/Season 1/Name (year) - S01E02.mkv
```
The issue is that all episodes get grouped as versions. I also tried
```
Name (2020)/Season 1/Name (year) S01E01.mkv
Name (2020)/Season 1/Name (year) S01E01 - [VERSION].mkv
Name (2020)/Season 1/Name (year) S01E02.mkv
```
with the same results. I tried making sense of the code itself, but frankly I do not quite understand the its logic.

Is there anything I'm doing wrong? Or should this work?

SenorSmartyPants commented:
I think you need all versions to have a name. Try this instead and let me know.

```
Name (2020)/Season 1/Name (2020) - S01E01 - [VERSION].mkv
Name (2020)/Season 1/Name (2020) - S01E01 - [VERSION2].mkv
Name (2020)/Season 1/Name (2020) - S01E02.mkv
```

SenorSmartyPants commented:
Did you merge all the commits in the PR not just the first one? There was an issue with all episodes merging before.

If my tip didn't help, can you include a screen shot of the episode all the videos are getting merged into with the version selection drop down open?

JeWe37 commented:
Yes, I merged all of them, I can't really test right now, I'll get back to this though.

JeWe37 commented:
@SenorSmartyPants I got around to testing it, even with everything named according to
Name (2020)/Season 1/Name (2020) - S01E01 - [ORIGINAL].mkv
Name (2020)/Season 1/Name (2020) - S01E01 - [VERSION].mkv
Name (2020)/Season 1/Name (2020) - S01E02 - [ORIGINAL].mkv
I still get the same effect of all episodes being shown as versions.

SenorSmartyPants commented:
I'll take a look at this today.

SenorSmartyPants commented:
@JeWe37 I was able to reproduce this with your Name (2020) example. So I will continue to dig into this.

It did work correctly when I used a different example to check if it was related to having the year in the series name.
```
Red/Season 1/Red - S01E01 - [ORIGINAL].mkv
Red/Season 1/Red - S01E01 - [VERSION].mkv
Red/Season 1/Red - S01E02 - [ORIGINAL].mkv
```

SenorSmartyPants commented:
@JeWe37 Lastest commit should fix your issue. Thanks for reporting it!

JeWe37 commented:
I'll give it a shot tomorrow, thanks for the quick resolution.

cvium commented:
I don't like that a folder is required. We need fewer filesystem rules, not more.

SenorSmartyPants commented:
> I don't like that a folder is required. We need fewer filesystem rules, not more.

A folder is no longer required. Other than the folder requirements that already exist, that is a series folder.

Fedec96 commented:
Any update on this? This would be big.

SenorSmartyPants commented:
> Any update on this? This would be big.

This needs to be reviewed and approved by team JF. Maybe I could get some feedback from @Bond-009 @Shadowghost @crobibero @cvium . I think they are some of the core server people.

dotbanana commented:
This is such an obvious feature i cannot understand why this wasn't implemented earlier and why this is taking so long now.

JeWe37 commented:
If it's any help: I've been running a jellyfin instance with this PR merged for a while now and it's been working fine. I have a fairly consistently named library, which is required to really get this feature to work, but I have not seen it interfering anywhere I did not want it to.

SenorSmartyPants commented:
> If it's any help: I've been running a jellyfin instance with this PR merged for a while now and it's been working fine. I have a fairly consistently named library, which is required to really get this feature to work, but I have not seen it interfering anywhere I did not want it to.

Awesome. Glad someone is getting some use out of it other than myself.

mammo0 commented:
Hello,
I'm also thinking about testing this PR.

Does my existing library get automatically merged by a simple rescan? Or is some manual work needed before?

Thank you for this PR :)

SenorSmartyPants commented:
> Hello, I'm also thinking about testing this PR.
> 
> Does my existing library get automatically merged by a simple rescan? Or is some manual work needed before?
> 
> Thank you for this PR :)

Episodes you want merged will have to follow the naming guidelines, but after that a rescan of those files should do it.

mammo0 commented:
> Episodes you want merged will have to follow the naming guidelines, but after that a rescan of those files should do it.

Works like a charm :)
Thank you.

TheMelmacian commented:
Can Team Jellyfin provide some feedback what's necessary to get this merged? This feature would be really helpful to cleanup my library. And based on the comments I'm not the only one that would benefit from this.

The same goes for https://github.com/jellyfin/jellyfin/pull/7997

Fedec96 commented:
Any chance this makes it for 10.9? @Bond-009 @Shadowghost  @crobibero @cvium

dotbanana commented:
I would also need some Tutorial how to do this because this ain't some scripting language, no? 
It wouldn't be enough to just exchange the lines of codes in the files since this is dotnet compiled stuff, right? 
I am not very well versed in these windows things.

anthonylavado commented:
> Also i wanted to ask if there is some problem going on with jellyfin. When i checking out the responsible folks most of them are inactive. The one person with the most committed lines has even stopped developing for jellyfin and is doing emby-things. Is this a dead horse? 

No, and no. Also, Jellyfin formed from Emby, that's why the person with the most lines is doing Emby things - that's where it all started.

asiaclonk commented:
Very interested in this as well. I just started migrating from Plex to Jellyfin and have a lot of trouble trying to merge versions of streams.

I have to go the library view and select "episodes" to view a list of all episodes and merge them from there, since there is no metadata for me to do it from the "More from this season" view.

Once merged, I have to set the metadata and images for each episode multiple times (the number of versions), because when displaying episode information, it seems to pick the .nfo from an arbitrary version. I hope this PR allows you to have a single .nfo and image for all versions of an episode.

SenorSmartyPants commented:
Latest commit will merge episodes if their cleaned filenames match.
This allows files like 
```
Name (2020) - S01E01 [BluRay-480p x264][AC3 2.0] - [ORIGINAL].mkv
Name (2020) - S01E01 [BluRay-1080p x264][AC3 5.1]- [Remaster].mkv
```
to be grouped on 'Name (2020) - S01E01'

Fedec96 commented:
Cool. Hopefully it gets merged in the next version.

SenorSmartyPants committed:
Make sure episodes in series folder don't group as multiversion


Test for episodes in series folder


Turn on multiple versions for episodes


Support for multiversion episodes in mixed folders

Update 2 failing test cases. These were for passing for unofficially supported filenames. Dashes with no spaces, which is not how JF docs say multiversion files are supposed to be named.
Fix possible null


fix null take 2


Don't ParseName when calling ResolveVideos<Episode> 

ParseName drops everything after year in a filename. This results in episode information being dropped if year is present.

Update tests to set ParseName=false

Additional test with Mixed folder with Year in filename

Added case when calculating displayname for versions for mixed folders.
Add StringComparison.Ordinal to LastIndexOf

Was generating an error in recent build attempts.
Clean the episode filename to set the grouper value

This allows files like 
Name (2020) - S01E01 [BluRay-480p x264][AC3 2.0] - [ORIGINAL].mkv
Name (2020) - S01E01 [BluRay-1080p x264][AC3 5.1]- [Remaster].mkv

to be grouped on 'Name (2020) - S01E01'
Fix false positive merging

Only do cleanstring or " -" index cleaning, not both.
Compatiblity fix when stacking episodes and multiple versions are present


Fix linting problems

SenorSmartyPants commented:
Force pushed to rebase and fix merge conflicts. Refactored some code in the process.

mammo0 commented:
Hello,
I found a bug that hit me after commit 270fe2c, especially the lines in the file `Emby.Naming/Video/VideoListResolver.cs`:

```c#
if (VideoResolver.TryCleanString(s, namingOptions, out var newName))
{
    s = newName;
}
```

I have the following file naming:
```
/TV/Name (2020)/Season 1/S01E01 - UHD.mkv
/TV/Name (2020)/Season 1/S01E01 - HD.mkv
/TV/Name (2020)/Season 1/S01E02 - HD.mkv
```

Normally the episode 1 should be grouped. But the above code uses the following regex to clean the file name (from `CleanStrings` property in file `Emby.Naming/Common/NamingOptions.cs`):
```regex
^\s*(?<cleaned>.+?)[ _\,\.\(\)\[\]\-](3d|sbs|tab|hsbs|htab|mvc|HDR|HDC|UHD|UltraHD|4k|ac3|dts|custom|dc|divx|divx5|dsr|dsrip|dutch|dvd|dvdrip|dvdscr|dvdscreener|screener|dvdivx|cam|fragment|fs|hdtv|hdrip|hdtvrip|internal|limited|multisubs|ntsc|ogg|ogm|pal|pdtv|proper|repack|rerip|retail|cd[1-9]|r3|r5|bd5|bd|se|svcd|swedish|german|read.nfo|nfofix|unrated|ws|telesync|ts|telecine|tc|brrip|bdrip|480p|480i|576p|576i|720p|720i|1080p|1080i|2160p|hrhd|hrhdtv|hddvd|bluray|blu-ray|x264|x265|h264|h265|xvid|xvidvd|xxx|www.www|AAC|DTS|\[.*\])([ _\,\.\(\)\[\]\-]|$)
```

As you can see this regex hits the first file (`S01E01 - UHD.mkv`). That means the name that is used for grouping is: `S01E01 - `.

For the second file (`S01E01 - HD.mkv`) none of the naming options regexes match. So the group name is processed with:
```c#
var idx = s.LastIndexOf(" -", StringComparison.Ordinal); // name doesn't come in with trailing space
s = idx == -1 ? s : s[..idx];
```
This results in `S01E01`, which means that the two versions of episode 1 will not get grouped.

A short solution would be to add the chars `HD` to the above naming options regex. After that the grouping works again for my naming scheme.
But I don't know if this will break something else.

SenorSmartyPants commented:
I just tested this with my updated squashed commit from yesterday (to resolve merge conflicts with the main branch) and it works as expected. I did change some of the logic in making the grouper strings.

Please try with the latest from this PR and let me know if this is still a problem.

I quickly dropped this test case into MultiVersionTest.cs
```c#
        [Fact]
        public void TestMultiVersionEpisodeMixedSeasonFolderWithNoSeriesName()
        {
            var files = new[]
            {
                @"/TV/Name (2020)/Season 1/S01E01 - UHD.mkv",
                @"/TV/Name (2020)/Season 1/S01E01 - HD.mkv",
                @"/TV/Name (2020)/Season 1/S01E02 - HD.mkv",
            };

            var result = VideoListResolver.Resolve(
                files.Select(i => VideoResolver.Resolve(i, false, _namingOptions, false)).OfType<VideoFileInfo>().ToList(),
                _namingOptions,
                CollectionType.TvShows);

            Assert.Equal(2, result.Count);
            Assert.Single(result[0].AlternateVersions);
            Assert.Empty(result[1].AlternateVersions);
        }
```

mammo0 commented:
Thanks for the fast response :)

I've seen that there are some differences between your branches. Unfortunately the `EpisodeMultipleVersions` can't be merged onto the release 10.8.9. Trying to figure it out...

SenorSmartyPants commented:
Yea, that's more of a problem to merge to 10.8.9. This PR target's main and not release-10.8.z

I privately have a version I run against 10.8.z, but I have not updated that with these latest changes.

SenorSmartyPants commented:
@mammo0 Trying merging from https://github.com/SenorSmartyPants/jellyfin/tree/10.8.z-EpisodeVersions . I just backported the PR changes and your test case.

mammo0 commented:
Thank you so much :)

Now everything works again.

Shadowghost commented on Emby.Naming/Video/VideoListResolver.cs:
```diff
@@ -129,35 +151,59 @@ private static List<VideoInfo> GetVideosGroupedByVersion(List<VideoInfo> videos,
+                    list.Add(OrganizeAlternateVersions(grouping.ToList(), grouping.Key.AsSpan(), primary));
+                }
+            }
+            else if (mergeable.Count() > 0)
```
```suggestion
            else if (mergeable.Count > 0)
```

Shadowghost replied:
It's also be nice if comment would start capitalized

SenorSmartyPants replied:
Implemented

SenorSmartyPants committed:
Code review: Count() -> Count and comments capitalization

Shadowghost commented:
Since I haven't seen tests for this layout but it's supported by Jellyfin - does it work with your changes (if it does, tests should be added for it too)?
```
SeriesName/SeriesName- S01E01 - EpisodeTitle - [VersionA].mkv
SeriesName/SeriesName- S01E01 - EpisodeTitle - [VersionB].mkv
SeriesName/SeriesName- S01E01 - EpisodeTitle - VersionC.mkv
SeriesName/SeriesName- S01E01 - EpisodeTitle - VersionD.mkv
```

mammo0 commented:
@Shadowghost I've tested your samples against the current implementation but the test fails.

After a little bit of debugging I saw that the files `SeriesName/SeriesName- S01E01 - EpisodeTitle - VersionC.mkv` and `SeriesName/SeriesName- S01E01 - EpisodeTitle - VersionD.mkv` are the culprit.

The file names hits the regex `^(?<filename>.*?)(?:(?<=[\]\)\}])|[ _.-]?)(?<number>[a-d])(?:\.[^.]+)?$` which is part of the `VideoFileStackingRules`:
https://github.com/jellyfin/jellyfin/blob/016cd22909a46edd6595f2b82ddec3e7001637db/Emby.Naming/Common/NamingOptions.cs#L141-L146

The regex is applied case insensitive, so it matches the `C` and the `D` after `...Version`. Because of that those two files are treated as parts not as versions.

I think this should not happen. According to the naming scheme in the docs (https://jellyfin.org/docs/general/server/media/shows/#episodes-split-across-multiple-parts) a `<parttype>` is required. But the above regex makes this optional.

mammo0 commented:
If I comment out the corresponding regex, the test case of @Shadowghost works.

But then the following test cases fail:

- Jellyfin.Naming.Tests.Video.StackTests.TestFalsePositive
- Jellyfin.Naming.Tests.Video.StackTests.TestMixedExpressions
- Jellyfin.Naming.Tests.Video.VideoListResolverTests.TestFourSisters

mammo0 commented:
I had a closer look at the failing tests:

1. `Jellyfin.Naming.Tests.Video.StackTests.TestFalsePositive`
Why should the files 
    - `300a`,
    - `300b` and
    - `300c`

    be treated as a single file? This behavior is not documented: https://jellyfin.org/docs/general/server/media/movies

2. `Jellyfin.Naming.Tests.Video.StackTests.TestMixedExpressions`
Same as 1.

3. `Jellyfin.Naming.Tests.Video.VideoListResolverTests.TestFourSisters`
The naming of the files
    - `Four Sisters and a Wedding - A.avi` and
    - `Four Sisters and a Wedding - B.avi`
    
    does rather match the multiple version naming scheme: https://jellyfin.org/docs/general/server/media/movies#multiple-versions-of-a-movie
    According to this scheme I expect to have two versions of the same movie (`A` and `B`) and not two parts.

Shadowghost commented:
I've the same sentiments than you. Those test don't do what the docs indicate they should. And in those specific cases I don't think the tests are correct.

mammo0 commented:
@Shadowghost I've created the PR #9496 that should solve this problem.

SenorSmartyPants commented:
Stacking takes place first in the existing code base. This code currently does not attempt to merge stacked files. That was causing issues for me earlier. But I could look into it, not sure if JF-web could handle that thou.

SenorSmartyPants commented:
> I don't think this works for movies and IMHO we supporting stacked versions is out of scope.

What doesn't work for movies?

Agreed, I wasn't worried about supporting multiple versions where some are stacked. Just stating that it explicitly doesn't.

Shadowghost commented:
Combining stacking + versioning doesn't work for movies was what I meant.

SenorSmartyPants committed:
version[ABCD] test

SenorSmartyPants commented:
Rebased to get Stacking fix update. Add new version test described by @Shadowghost .

captainnapalm commented:
As I understand this PR, it requires the episodes to be in the same directory? So if I'm maintaining 1080p and 4k versions using separate instances of Sonarr, which require separate roots, this PR will not help to address this. I currently have 2 directories mapped to my TV Shows library in Jellyfin. I'm using the Merge Version plugin but that in and of itself isn't 100% foolproof.

SenorSmartyPants commented:
> As I understand this PR, it requires the episodes to be in the same directory? So if I'm maintaining 1080p and 4k versions using separate instances of Sonarr, which require separate roots, this PR will not help to address this. I currently have 2 directories mapped to my TV Shows library in Jellyfin. I'm using the Merge Version plugin but that in and of itself isn't 100% foolproof.

Yes, episodes must be in the same directory, or appear that way to jellyfin. You could use something like mergerfs to create a merged view of your TV content for jellyfin.

JeWe37 commented:
Hey @SenorSmartyPants, are you still working on this? I have a rebased version that I made for my own use, which was relatively easy to make for now, but it's only going to get harder as time goes on without maintenance. I really would like to finally get this merged such that I can stop maintaining a private fork.

SenorSmartyPants commented:
Yes, I have not abandoned this. Personal life has gotten in the way of fixing the merge conflicts that show up occasionally since this has sat so long without merging.  I will fix this when things slow down for me a little.

Deedikjupijn commented:
wow already a year waiting for this, hope this gets merged soon!

SenorSmartyPants commented:
Rebase, fix merge conflict and force push.

SenorSmartyPants committed:
Convert Equals 1 to Single tests

adityaruplaha commented:
Would love to have this, is there anything blocking this?

SenorSmartyPants commented:
> Would love to have this, is there anything blocking this?

This needs another approval from team JF. Maybe @Shadowghost can request another team member to review. I don't seem to be able to do that.

Everyone else that likes this feature please upvote with emoji on the first comment.

cvium commented:
I am concerned about the performance impact of this PR. There's many lists being created, filenames are parsed multiple times and EpisodeResolver is being created in a loop etc.

Design-wise, I think we should re-think the flow. I don't think it makes sense to do the grouping until after all files have been parsed and metadata has been fetched. Relying on filenames is brittle.

SenorSmartyPants commented:
> I am concerned about the performance impact of this PR. There's many lists being created, filenames are parsed multiple times and EpisodeResolver is being created in a loop etc.

Are there any tools/tests that are currently being used to assess JF performance? I could add that tooling to see the performance impacts.

> Design-wise, I think we should re-think the flow. I don't think it makes sense to do the grouping until after all files have been parsed and metadata has been fetched. Relying on filenames is brittle.

Your issue with the flow and reliance on filenames seems like an issue with the underlying architecture of JF and not my PR. Respectfully, please start a discussion about that elsewhere.

exterrestris commented:
The documentation states that multiple versions for movies are defined using common filenames, not metadata - I can't see a sensible reason why episodes would be any different.

In my experience of multiple movie versions, the metadata fetcher isn't reliable enough to be used for grouping files - I have plenty of multi-version movies that are named according to the documentation, but Jellyfin gets the metadata wrong and doesn't group them

cvium commented:
> > I am concerned about the performance impact of this PR. There's many lists being created, filenames are parsed multiple times and EpisodeResolver is being created in a loop etc.
> 
> Are there any tools/tests that are currently being used to assess JF performance? I could add that tooling to see the performance impacts.

JetBrains dotMemory, dotTrace. ANTS Memory Profiler. dotnet monitor. The tools are many and varied, but ultimately nearly identical. It doesn't really require any tools to see the potential issue here though. For many people it will not have a noticeable impact, but any performance issues within the scanner scales directly with the number of files in people's libraries.

- More memory being used means more garbage to collect.
- More garbage to collect means longer and more frequent GC time.
- More time spent in the scanner potentially means longer database locks, which makes the UI slower.
- Etc.

> > Design-wise, I think we should re-think the flow. I don't think it makes sense to do the grouping until after all files have been parsed and metadata has been fetched. Relying on filenames is brittle.
> 
> Your issue with the flow and reliance on filenames seems like an issue with the underlying architecture of JF and not my PR. Respectfully, please start a discussion about that elsewhere.

My issue is with the design of JF, but your change has now coupled VideoListResolver with EpisodeResolver to achieve your goal. There's already too much coupling in general, I don't like to add more.

JeWe37 commented:
Frankly, to me this seems like its blocking a well requested feature solely based on a broader, out-of-scope architectural issue. Now, I understand that you can't neglect the users requiring the performance, even if for this to be drastic you'll likely need a library in the 100s of TB. I think a fair compromise here would be to simply make the merging configurable(both for movies and tv shows). This would also satisfy the people worried about overly eager merging. Just put it in an easy to revert commit that can be removed whenever the redesign of the broader architecture is complete.

JeWe37 commented:
Alright decided that I might as well make it configurable myself. See JeWe37/jellyfin-web@654d4418e0c63073c58d15b56c466a84d3af22ab for the jellyfin-web changes and JeWe37/jellyfin@e585c6c863758eb74a2c52ed167c3f629bc05e38 for the jellyfin changes.

I would appreciate if @cvium could give us his OK for this with it being entirely optional now. This should alleviate all of his concerns.

After we have that confirmed, it might also be good if you @SenorSmartyPants pulled in that commit to your branch, so no new PR is needed here. Once that is done I can also open the corresponding jellyfin-web PR.

I'd love to see the approvals coming together for this such that it can be merged before 10.9.

mammo0 commented:
Hello,

for anyone that is interested:
I created a Docker image with the changes of this PR: https://hub.docker.com/r/mammo0/jellyfin_multi-episode-versions

It is based on the (currently) latest 10.9.1 release and I only applied the changes of @SenorSmartyPants and @JeWe37. So, if you want to use this feature, you have to enable it in the settings page of each library.

I will try to keep it up to date... (Until this PR is merged and is part of an official Jellyfin release)

jhparker88 commented:
> My issue is with the design of JF, but your change has now coupled VideoListResolver with EpisodeResolver to achieve your goal. There's already too much coupling in general, I don't like to add more.

I understand the technical hesitation but the fact remains the current situation is not user friendly at all for the user who watches and the person who maintains the library.

Again, I also like to see this enhancement to be available because right now the handling of episodes that have multiple versions isn't appealing at all and creating different folders to hold different versions for the same show doesn't seems to be the right choice given that the semantics of show X with episode SE01E01 being identical for the following versions SE01E01 - 540p, SE01E01 - 720p, and SE01E01 - 1080p hence representing the same entity (not three different once).

If one can enable it on a per library basis then there shouldn't be much of a performance problem given that a user has to deliberately enable it and understand the implications of it.

mammo0 commented:
I fixed the current merge conflict like this: https://github.com/mammo0/jellyfin/commit/ca1f0d7b295fbc8fa90fc069427dd756df9c2280#diff-bcc2125e56d5738b4778802ac650ca47719845aeee582f3b5c9b46af82ea9979

@SenorSmartyPants It passes all tests. So it should work.

mammo0 commented:
Hello,

I updated my fork to the latest 10.10.0 release: https://github.com/jellyfin/jellyfin/compare/master...mammo0:jellyfin:10.10.z-mammo0
There were no merge conflicts since the last 10.9.11 release. Only a small fix for a test case was needed: https://github.com/jellyfin/jellyfin/commit/f4f0051485f5ddc7339a15cd37749e71ead9ce0d

**All test cases are passing :)**

I also updated my docker images on: https://hub.docker.com/r/mammo0/jellyfin_multi-episode-versions

craftablescience commented:
This feature would be huge, I can't believe I have to switch to a fork to use it lmao. Thank you so much for making this PR

It's been two and a half years 😭

qwertyuioasdfghjjkllmnbcvz commented:
> **Changes** Enable multiple versions for episodes. Episode suffixes work or episode may be placed in it's own directory. Will not collapse episodes in series directory. Unit tests included.
> 
> ```
> /TV/Dexter/Dexter - S01E01/Dexter - S01E01 - One.mkv
> /TV/Dexter/Dexter - S01E01/Dexter - S01E01 - Two.mkv
> /TV/Dexter/Dexter - S01E02 - Another.mkv
> /TV/Dexter/Dexter - S01E03 - Again.mkv
> /TV/Dexter/Season 2/Dexter - S02E01/Dexter - S02E01 - First.mkv
> /TV/Dexter/Season 2/Dexter - S02E01/Dexter - S02E01 - Second.mkv
> /TV/Dexter/Season 2/Dexter - S02E02 - More.mkv
> ```
> 
> JF will show 3 episodes in season 1. 2 episodes in season 2.

I would include cross over episodes like murder she wrote the Andy Griffith show gomer Pyle etc that way it pulls from one file

JPVenson commented on Emby.Naming/Video/VideoListResolver.cs:
```diff
@@ -207,5 +261,41 @@ private static bool IsEligibleForMultiVersion(ReadOnlySpan<char> folderName, Rea
+        {
+            // Grouper for tv shows/episodes should be everything before space-dash-space
+            var resolver = new EpisodeResolver(namingOptions);
+            EpisodeInfo? episodeInfo = resolver.Resolve(testFilePath, false);
```
use var

mammo0 replied:
did you mean to NOT use `var`?

mammo0 replied:
Sorry, I don't get what you mean...

JPVenson replied:
```suggestion
            var episodeInfo = resolver.Resolve(testFilePath, false);
```

mammo0 replied:
Ah OK :)

I updated my PR accordingly (https://github.com/SenorSmartyPants/jellyfin/pull/121/commits/2e702fb1ac222e3eea10be8d60042a43353df9fe)

SenorSmartyPants replied:
I updated this PR with that change.

JPVenson reviewed (changes requested):
minor coding style detail, otherwise LGTM

mammo0 commented:
@JPVenson: What about the changes from @JeWe37?
With them the grouping of movies and TV shows can be enabled/disabled via the web interface.

JeWe37 commented:
So far @SenorSmartyPants hasn't implemented them, though it might be best to avoid a new PR so that the approvals aren't needed again. Can one of the maintainers perhaps push those directly? Or the rebased version from @mammo0's fork anyway.

SenorSmartyPants commented:
I'm hesitant to put effort into this without some assurances it will actually get merged. It's been a long time and my current energy level for projects is pretty low. Or if a JF team member wants to make alterations to get this merged you have my permission.

mammo0 commented:
@SenorSmartyPants If you want I can continue this PR. Since we have already an approval from @Shadowghost and now a 'LGTM' (after a small change) from @JPVenson, I hope this will get merged soon.

@SenorSmartyPants Can you give me write access to your fork (especially the `EpisodeMultipleVersions` branch)? Then I should be able to push changes directly to this PR.
Otherwise I can open a new PR and link it to this one.

What option is preferred?

@JeWe37: I can also open a PR on the `jellyfin-web` repo with your changes. Or you do it on your own.

SenorSmartyPants commented:
> Can you give me write access to your fork 

@mammo0 I don't see any easy way to give you access to just one branch on my repo. So instead, could you make a PR to https://github.com/SenorSmartyPants/jellyfin/tree/EpisodeMultipleVersions which I can then merge and that will update this PR?

mammo0 commented:
> @mammo0 I don't see any easy way to give you access to just one branch on my repo. So instead, could you make a PR to https://github.com/SenorSmartyPants/jellyfin/tree/EpisodeMultipleVersions which I can then merge and that will update this PR?

Please have a look at https://github.com/SenorSmartyPants/jellyfin/pull/121 :)

SenorSmartyPants committed:
Update tests

SenorSmartyPants commented:
I fixed merge conflicts and tests. I did not include the library option to enable/disable grouping. Let that be a separate PR after this is merged. So this PR does not require any web changes.

gnattu commented:
This is a bad time to implement functions like this because we are in the middle of a huge database refactoring, and we may need to design a better data structure for multi version videos so that it does not cause bugs everywhere, especially for playlists and collections. The already merged movie versions has being a maintenance headache and I don’t want to see episodes to be in the same situation.

I think it’s best to hold off on adding these functions right now. We’re currently in the middle of a major database refactoring, and we need to make sure we have a solid data structure for multi-version videos before we proceed. We don’t want to end up with a bunch of bugs, especially with playlists and collections like the already merged muiltversion movie, which has been a pain to maintain, and I don’t want to see episodes get caught up in the same mess. Let’s wait until things are a bit more stable before we dive into this.

fishcharlie commented:
@gnattu Totally reasonable. Do you think after that database migration this can be looked at again? This would be a super useful feature to have in Jellyfin. I really hope it doesn't get abandoned forever.

JeWe37 commented:
There are no database changes with this since it only makes it so that episode merging happens automatically rather than manually. While this could exacerbate any possible issues by making merged series more common, even in the worst case of an imperfect database migration in the future, a rescan should deal with any merging-related issues with little hassle.

Frankly it is somewhat frustrating to hear that this still cannot move forward. For my part I have been running a fork of jellyfin because of this(and similar issues) for multiple years now and I seem to not be the only one frustrated with the current situation:

> I'm hesitant to put effort into this without some assurances it will actually get merged. It's been a long time and my current energy level for projects is pretty low. Or if a JF team member wants to make alterations to get this merged you have my permission.

This is particularly true with this now being one of the [more requested, yet still not available feature requests](https://features.jellyfin.org/posts/251/multiple-versions-drop-down-on-tv-series).

gnattu commented:
> There are no database changes with this since it only makes it so that episode merging happens automatically rather than manually. While this could exacerbate any possible issues by making merged series more common, even in the worst case of an imperfect database migration in the future, a rescan should deal with any merging-related issues with little hassle.

If writing migrations and doing data schema refactors sounds easy to you can you help us with the current refactor so that we can move to the right direction faster? https://github.com/jellyfin/jellyfin/pull/12798

> Frankly it is somewhat frustrating to hear that this still cannot move forward. For my part I have been running a fork of jellyfin because of this(and similar issues) for multiple years now and I seem to not be the only one frustrated with the current situation:

I am also frustrated by current multi version implementation as it is causing bugs due to the design flaw. It failed to predict a lot of common use cases and it is hard to fix because the already written data is hard to migrate even if we come up with a better design. I want to have this anticipated feature to land just like everyone in the feature request ticket because I want to implement features based on this. I just want to make sure the design flaws are addressed before we can adding more features like this.

JeWe37 commented:
> If writing migrations and doing data schema refactors sounds easy to you can you help us with the current refactor so that we can move to the right direction faster? #12798

My point is solely about the fact that merging this will not make the database issues worse. I obviously cannot claim that I were intimately familiar with the issues around multiple versions, though for the past years it has been working essentially fine for me. 

Now, of course this doesn't mean those issues don't exist, but it does mean this PR can be used productively and, since it does not introduce anything new to the database format, inherently cannot change or complicate anything about the migrations that already have to be written.

JPVenson commented:
whats not that widely known is that the current refactor brought up so many many issues with the codebase that streach into all areas of the code. The database is the core of jellyfin and unfortunatly a lot of bad design choices were made based on the horrific database design.

Now we can't just replace the database with better code, that would be far to easy and would have been done many years ago. To make the database _somewhat_ usable, the code needs to still support attrocious design choices, like querying the database 1+n² times just for the next-up query. Sure we can rewrite that but i hope you see where i am going with that. I was so close to rewriting even bigger chunks just to get a somewhat sensible UI structure and the Pr is already +20.000 lines of code.

Overall we are currently focusing hard on the EfCore rewrite, because it has to be done _some time_ and that time is now, which on the other hand means that we cannot really bring much else in right now because we just dont have the capacity for it. Please dont underestimate the reach the EfCore PR has, it touches like 60-70% of the codebase and adding even more new stuff to that will make it borderline impossible to debug in the end.

DLLarson commented:
Please make this feature available. I have episodes of several series in both 1080p and DVD resolutions and the only way I can seem to do it is by manual grouping. That's fine but I can't get them sort the 1080p entry first in the list no matter how I select the order.

It's my hope that this feature will solve my problems. 

I see this feature request pop up from time-to-time on Reddit but all the suggested workarounds have either been ineffective or even hard to understand like the statement "edit the metadata to control the sorting order." HOW??? there is no sort option.

maxdorninger commented:
Is there an ETA for this feature?

Speiger commented:
@SenorSmartyPants 
Is there the possiblity of doing this for example:
```
/TV/Dexter/S01/Dexter - S01E01
/TV/Dexter/S01 - YT/Dexter - S01E01
/TV/Dexter/S02/...
/TV/Dexter/S03/...
```

Where you still need to store them in the same master folder but can have dedicated folders for dedicated versions?
The idea being that you have to change less file names and allow the merging to be done on a folder basis?

Not sure how viable that is but that would be epic :)

LightTemplar commented:
@Speiger
> Not sure how viable that is

I have a use case for exact this: for one show I've got a simple "season 1" and "season 1 director's cut".

IngwiePhoenix commented:
Came here after dumping Interspecies Reviewers from the boxset I bought at a convention. The set comes with _both_ the Blu-Ray (1080p) and DVD (720p) versions, where each volume has a Blu-Ray disc and two DVDs with the corresponding episodes in that volume - and I have both. Having the same episode appear twice is a little frustrating... Would love to see a workaround. Because, I honestly have no idea if those episodes genuenly are different or not, so I would like to keep both, just in case.
I had also thought of just merging both versions into one MKV; not sure if that would be a "solution" or a "hack" though :p

Speiger commented:
@IngwiePhoenix Usually the only difference between DVD & Bluray versions are 2 things.
Subtitles and Video Quality.
Subtitles on blurays are images and can't be changed. While on DVD their either .srt or .ass (Yes that's a real format) files
And personally i like .srt/.ass files more mainly because they are rendered on the fly vs images that are scaled.

and video quality is well, BluRays are 1080p with the bitrate of 4k, while dvd is 720p with a bitrate of 1080p.
So maybe merge the video & audio of the bluray and the subtitles of the dvd.
There are ffmpeg wrappers out there that can do it, or write your own (what i have done)

If you want to keep both versions, you can solve that really easily too, but that is a hack more than a solution.
Create 2 Libraries directing to 2 different folders and both dump them into each folder.
Jellyfin will identify them both and keep them both split while the episode tracker will keep track of them both combined.
So if you watch a ep of the dvd version it will also update the BluRay version as being watched.

Hope that helps

jwadamson commented:
NTSC DVDs are 480p (720x480); this is NOT the same as 720p but still was
amazing compared to analog 480i that VHS and broadcast used. Resolutions
prior to 4k (3840x2160) used their exact height resolution with either a i
or p suffix as the short-name.

Have you tried enabling the "Merge Versions"  plugin?
https://github.com/danieladov/jellyfin-plugin-mergeversions

I have multiple versions of TV episodes and they are managed as a single
entry. The individual episode screen has a "version" dropdown to pick
between the file names despite files being in different subfolders of the
show directory e.g.

./{show}/MLB 2015 Season 5 WEB-DL 720p x264 [i_c]/MLB - S05E02 -
{eptitle}.mkv
./{show}/mlb..s05e02.{eptitle}.mp4

--Jeff


On Sun, Apr 20, 2025 at 1:24 AM Ingwie Phoenix ***@***.***>
wrote:

> Came here after dumping Interspecies Reviewers from the boxset I bought at
> a convention. The set comes with *both* the Blu-Ray (1080p) and DVD
> (720p) versions, where each volume has a Blu-Ray disc and two DVDs with the
> corresponding episodes in that volume - and I have both. Having the same
> episode appear twice is a little frustrating... Would love to see a
> workaround. Because, I honestly have no idea if those episodes genuenly are
> different or not, so I would like to keep both, just in case.
> I had also thought of just merging both versions into one MKV; not sure if
> that would be a "solution" or a "hack" though :p
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jellyfin/jellyfin/pull/8004#issuecomment-2817000530>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABUOK7BIYA266M43AH6GLD22MVRDAVCNFSM5ZZI7SV2U5DIOJSWCZC7NNSXTN2JONZXKZKDN5WW2ZLOOQ5TEOBRG4YDAMBVGMYA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
> *IngwiePhoenix* left a comment (jellyfin/jellyfin#8004)
> <https://github.com/jellyfin/jellyfin/pull/8004#issuecomment-2817000530>
>
> Came here after dumping Interspecies Reviewers from the boxset I bought at
> a convention. The set comes with *both* the Blu-Ray (1080p) and DVD
> (720p) versions, where each volume has a Blu-Ray disc and two DVDs with the
> corresponding episodes in that volume - and I have both. Having the same
> episode appear twice is a little frustrating... Would love to see a
> workaround. Because, I honestly have no idea if those episodes genuenly are
> different or not, so I would like to keep both, just in case.
> I had also thought of just merging both versions into one MKV; not sure if
> that would be a "solution" or a "hack" though :p
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jellyfin/jellyfin/pull/8004#issuecomment-2817000530>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABUOK7BIYA266M43AH6GLD22MVRDAVCNFSM5ZZI7SV2U5DIOJSWCZC7NNSXTN2JONZXKZKDN5WW2ZLOOQ5TEOBRG4YDAMBVGMYA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

jhparker88 commented:
Could we please limit the discussion to the PR and its actual implementation details and avoid diverting into issue like "... NTSC DVDs ..." or "... difference between DVD & Bluray versions...". There is a forum to answer and respond to those issues hereby avoiding diluting the PR's content which is about the technical handling of "Episode multiple versions".

adityaruplaha commented:
> Have you tried enabling the "Merge Versions" plugin? https://github.com/danieladov/jellyfin-plugin-mergeversions

Thanks! This is exactly what I needed.

Speiger commented:
Eh, why was my first comment marked off topic?
It was asking about a detail about the implementation itself? And if it wasn't part of the implementation that it should maybe be added.

So kindly ask that the first comment i made asking for a implementation detail be please marked as on topic.
Ty

SenorSmartyPants commented:
I don't plan on working on this until multiple JF team members tell me that will be ready to review and probably merge almost immediately after I resolve conflicts. And then it will still depend on what's going on in my life at the time. I haven't abandoned this.

felix920506 commented:
@SenorSmartyPants I was told by @JPVenson that if people test it (needs merge conflicts solved first) and it works, he's willing to give a thumbs up

Exagone313 commented on Emby.Naming/Video/VideoListResolver.cs:
```diff
@@ -209,5 +264,41 @@ private static bool IsEligibleForMultiVersion(ReadOnlySpan<char> folderName, Rea
+            {
+                var rule = namingOptions.VideoVersionRegexes[i];
+                var match = rule.Match(filename);
+                if (!match.Success)
+                {
+                    continue;
```
(not a Jellyfin developer)

It is possible that more than one regex will exist in the future, but the current code is still looping over next regexes even if it matched. Why not rewriting it like this:

```cs
if (match.Success)
{
    g = match.Groups["filename"].Value;
    // ...
    // insert the rest of the code below
    // ...
    break;
}
```

An alternative would be use this form:

```cs
for (var i = 0; i < namingOptions.VideoVersionRegexes.Length; i++)
{
    var rule = namingOptions.VideoVersionRegexes[i];
    var match = rule.Match(filename);
    if (match.Success)
    {
        string g = match.Groups["filename"].Value;
        // ...
        // rest of the code
        // ...
        return g;
    }
}
return filename;
```

Exagone313 commented on tests/Jellyfin.Naming.Tests/Video/VideoListResolverTests.cs:
```diff
@@ -65,6 +67,87 @@ public void TestStackAndExtras()
+        {
+            var files = new[]
+            {
+                @"/TV/Grey's Anatomy (2005)/Grey's Anatomy (2005) - s01e01 CD1.avi",
+                @"/TV/Grey's Anatomy (2005)/Grey's Anatomy (2005) - s01e01 CD2.avi",
```
(not a Jellyfin developer)

Can you explain how these two end up grouped?

sic6SaNdMaN commented:
Is there anything in the current version of Jellyfin that works "out of the box" - even if there is a little workaround needed?

I have some series in multiple versions (some are upscaled and I would like to keep the originals, too, some are in different languages and not muxable, and so on, doesn't really matter).
What I would like to achieve is one of the following:

- Two separate entries for the whole series, so that I can rename them on my own ["BestSeries (original DVD)" + "BestSeries (upscaled)"]
- Two separate entries for each season (even if there's only one), so that I can rename them on my own ["Season 1 (original DVD)" + "Season 1(upscaled)"]

What I don't want to have (to do) is double entries of each episode. But that's what I currently get, if I try to have more then one version of any episode.
But it's definitely too much work to edit the metadata for each episode.

For movies I use the recommended way and it works like a charm (except for Kodi compatibility, but that's another story):
https://jellyfin.org/docs/general/server/media/movies#multiple-versions

So... what is possible for series?
Any suggestions are welcome.

Exagone313 commented:
Please be reminded that open source software is not entitled to you, it comes without warranty of any kind, as per the license terms. No need for hash words.

On the other hand, I am pretty sure that this plugin's maintainer would be happy if you contribute to its documentation.

Speiger commented:
@Exagone313 Yet the maintainers actively put stones infront of the developer who tries to contribute.
That invalidates your Argument entirely

sic6SaNdMaN commented:
@Seralpa: Thank you - that's a perfectly good explanation.
Please put it in the docs of the plugin ;) (all jokes aside, but it would be really helpful).

Wenn I will give it a try then - thanks again for the suggestion.

@admins: You can mark this as off-topic now.

dan-riley commented:
What's needed to get this merged? Anything I can do to help?
